### PR TITLE
Adding the feature to support importing and exporting API Products

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Support importing and exporting API Products from API Controller
Fixes: https://github.com/wso2/product-apim-tooling/issues/291

## Goals
Design and implement the REST APIs and functions in APIM as a feature, to support importing and exporting API Products (along with their dependent APIs) to the APIM. (Mainly this will be used from the API Controller)

## Approach
- Generated the WSO2 Admin Portal REST API (v1).
resized
- Introduced a new scope as **apim:api_product_import_export** to invoke the above REST APIs. (To use the above resources the user need to have **apim:api_product_import_export** scope. But note that, to use these resources in the scenarios where the user wants to import/update dependent APIs, the user should have not only the apim:api_product_import_export scope but also should have **apim:api_import_export** scope as well)
- Wrote new functions to import and export API Products in utils.
- Restructured the existing functions that have been used when importing and exporting APIs by refactoring the common methods that have been used in both import/export APIs and API Products, by creating appropriate classes.

Refer **[[APIM] Support for API Products from API Controller](http://wso2-oxygen-tank.10903.n7.nabble.com/APIM-Support-for-API-Products-from-API-Controller-td161499.html)** for further information.

## User stories
- Export an API Product with the meta-information, documents, images (thumbnails), client certificates and dependent APIs.
- The following main scenarios can be captured by the import API Product functionality.
  - Import a fresh API Product with a fresh set of dependent APIs*.
  - Import a fresh API Product when dependent APIs are already imported to APIM successfully and you do not want to update those APIs.
  - Import a fresh API Product when dependent APIs are already imported to APIM successfully and you want to update those APIs*.
  - Update the API Product only by changing the meta information and by adding/removing the resources of the API Product.
  - Update the API Product by adding new resources to both the API Product and the API(s)*.
  - Update only the dependent APIs*.
  - Import the API Product and its dependent APIs to another tenant (with isDefaultProviderAllowed=false)
  - Update only an already imported API Product and but not its dependent APIs in another tenant  (with isDefaultProviderAllowed=false)
  - Update an already imported API Product and its dependent APIs in another tenant (with isDefaultProviderAllowed=false)

(Note that, the scenarios which contain * mark are not allowed to be performed by a user who has only apim:api_product_import_export scope. The user need to have apim:api_import_export as well to perform those tasks)

## Release note
WSO2 Admin REST API supports importing and exporting API Products now.

## Documentation
Documentation changes should be done. 
https://github.com/wso2/product-apim-tooling/issues/288

## Related PRs
https://github.com/wso2/product-apim-tooling/pull/306

## Test environment
- JDK 1.8.0_241
- Ubuntu 18.04.4 LTS

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/ApiTypeWrapper.java [hunk 1]
```diff
@@ -68,4 +68,68 @@ public void setTier(String tier) {
...
+            api.setContextTemplate(contextTemplate);
+        }
+    }
+
+    public String getContextTemplate() {
+        if (isAPIProduct) {
+            return apiProduct.getContextTemplate();
+        } else {
+            return api.getContextTemplate();
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/ApiTypeWrapper.java:101]: Shall we use a ternary operator here?
```suggestion
        return isAPIProduct ? apiProduct.getContextTemplate() : api.getContextTemplate();
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -7380,19 +7420,22 @@ protected void checkAccessControlPermission(Identifier identifier) throws APIMan
...
-        // list to hold resouces which are actually in an existing api. If user has created an API product with invalid
+        // list to hold resources which are actually in an existing api. If user has created an API product with invalid
         // API or invalid resource of a valid API, that content will be removed .validResources array will have only
         // legitimate apis
         List<APIProductResource> validResources = new ArrayList<APIProductResource>();
         for (APIProductResource apiProductResource : resources) {
             API api = null;
-            try {
+            APIProductIdentifier productIdentifier = apiProductResource.getProductIdentifier();
+            if (productIdentifier != null) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:7430]:
```suggestion
            if (apiProductResource.getProductIdentifier() != null) {
```

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/ApiTypeWrapper.java (hunk 1 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/ApiTypeWrapper.java:101]: Sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 3]
```diff
@@ -808,6 +808,19 @@ public boolean isAPIAvailable(APIIdentifier identifier) throws APIManagementExce
...
+    public boolean isAPIProductAvailable(APIProductIdentifier identifier) throws APIManagementException {
+        String path = APIConstants.API_ROOT_LOCATION + RegistryConstants.PATH_SEPARATOR +
+                identifier.getProviderName() + RegistryConstants.PATH_SEPARATOR +
+                identifier.getName() + RegistryConstants.PATH_SEPARATOR + identifier.getVersion();
+        try {
+            return registry.resourceExists(path);
+        } catch (RegistryException e) {
+            String msg = "Failed to check availability of API Product :" + path;
+            log.error(msg, e);
+            throw new APIManagementException(msg, e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:820]: We shouldn't log and throw.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 4]
```diff
@@ -808,6 +808,19 @@ public boolean isAPIAvailable(APIIdentifier identifier) throws APIManagementExce
...
 
+    public boolean isAPIProductAvailable(APIProductIdentifier identifier) throws APIManagementException {
+        String path = APIConstants.API_ROOT_LOCATION + RegistryConstants.PATH_SEPARATOR +
+                identifier.getProviderName() + RegistryConstants.PATH_SEPARATOR +
+                identifier.getName() + RegistryConstants.PATH_SEPARATOR + identifier.getVersion();
+        try {
+            return registry.resourceExists(path);
+        } catch (RegistryException e) {
+            String msg = "Failed to check availability of API Product :" + path;
+            log.error(msg, e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:819]: Let's remove log.error. We shouldn't log and throw which is an antipattern.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/APIImportExportManager.java [hunk 5]
```diff
@@ -59,19 +64,25 @@ public APIProvider getApiProvider() {
...
+    public File exportAPIOrAPIProductArchive(ApiTypeWrapper apiTypeWrapper, boolean isStatusPreserved, ExportFormat exportFormat)
             throws APIImportExportException {
-
-        APIIdentifier apiIdentifier = apiToReturn.getId();
-        String archiveBasePath = exportAPIArtifacts(apiToReturn, isStatusPreserved, exportFormat);
+        String archiveBasePath;
+        if (!apiTypeWrapper.isAPIProduct()) {
+            APIIdentifier apiIdentifier = apiTypeWrapper.getApi().getId();
+            archiveBasePath = exportAPIArtifacts(apiTypeWrapper.getApi(), isStatusPreserved, exportFormat);
+            log.info("API" + apiIdentifier.getApiName() + "-" + apiIdentifier.getVersion() + " exported successfully");
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/APIImportExportManager.java:79]: Do we need this as an info log? Shall we make this "debug"?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+            throws APIImportExportException {
+
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        String docDirectoryPath = File.separator + APIImportExportConstants.DOCUMENT_DIRECTORY;
+        CommonUtil.createDirectory(archivePath + docDirectoryPath);
+        try {
+            for (Documentation doc : docList) {
+                String sourceType = doc.getSourceType().name();
+                String resourcePath = null;
+                String localFilePath;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:159]: Can we move this to Line no:183? 
```
String localFilePath = localDocDirectoryPath + File.separator + localFileName;
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+                String localFileName = null;
+                String localDocDirectoryPath = docDirectoryPath;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:161]: docDirectoryPath is not used further. So do we need a new variable? Can we reuse docDirectoryPath?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+    public static void exportClientCertificates(String archivePath, ApiTypeWrapper apiTypeWrapper, int tenantId, APIProvider provider,
+                                                 ExportFormat exportFormat) throws APIImportExportException {
+
+        List<ClientCertificateDTO> certificateMetadataDTOS;
+        try {
+            if (!apiTypeWrapper.isAPIProduct()) {
+                certificateMetadataDTOS = provider.searchClientCertificates(tenantId, null, apiTypeWrapper.getApi().getId());
+            } else {
+                certificateMetadataDTOS = provider.searchClientCertificates(tenantId, null, apiTypeWrapper.getApiProduct().getId());
+            }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:273]: Remove ! and swap the blocks

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+            if (!certificateMetadataDTOS.isEmpty()) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:274]: can change this as 
```suggestion
            if (certificateMetadataDTOS.isEmpty()) {
                return;
            }
            <rest of the code>
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+                Gson gson = new GsonBuilder().setPrettyPrinting().create();
+                String element = gson.toJson(certificateMetadataDTOS,
+                        new TypeToken<ArrayList<ClientCertificateDTO>>() {
+                        }.getType());
+
+                switch (exportFormat) {
+                    case YAML:
+                        String yaml = CommonUtil.jsonToYaml(element);
+                        CommonUtil.writeFile(archivePath + APIImportExportConstants.YAML_CLIENT_CERTIFICATE_FILE,
+                                yaml);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:286]:
```suggestion
                        CommonUtil.writeFile(archivePath + APIImportExportConstants.YAML_CLIENT_CERTIFICATE_FILE,
                                CommonUtil.jsonToYaml(element));
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 4 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:819]: Sure, the same procedure has been done in isAPIAvailable function in line numbers 805-807. Shall I change that too?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 3 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:820]: Same as above.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/APIImportExportManager.java (hunk 5 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/APIImportExportManager.java:79]: Sure will change it here and in line 83 as well.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java (hunk 6 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:159]: Sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java (hunk 7 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:161]: Yes we can resue it. Thanks for pointing it out.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java (hunk 8 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:273]: Sure, will swap where applicable.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java (hunk 9 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:274]: Sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java (hunk 10 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/utils/APIAndAPIProductCommonUtil.java:286]: Sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 2 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:7430]: Sure

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 4 above)
Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:819]: Yes please, let's do this for other applicable places too,

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 11]
```diff
@@ -5322,6 +5324,36 @@ private void saveAPIDefinition(APIProduct apiProduct, String apiDefinitionJSON,
         }
     }
 
+    @Override
+    public void addAPIProductSwagger(Map<API, List<APIProductResource>> apiToProductResourceMapping, APIProduct apiProduct)
+            throws APIManagementException {
+        APIDefinition parser = new OAS3Parser();
+        SwaggerData swaggerData = new SwaggerData(apiProduct);
+        String apiProductSwagger = parser.generateAPIDefinition(swaggerData);
+
```
Comment (vithu30) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:5333]: Shall we remove this empty line? Please check other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManager.java [hunk 12]
```diff
@@ -22,6 +22,7 @@
 import org.wso2.carbon.apimgt.api.dto.CertificateMetadataDTO;
 import org.wso2.carbon.apimgt.api.dto.ClientCertificateDTO;
 import org.wso2.carbon.apimgt.api.model.APIIdentifier;
+import org.wso2.carbon.apimgt.api.model.APIProductIdentifier;
```
Comment (vithu30) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManager.java:25]: Seems this import is not used. Shall we remove it?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java [hunk 13]
```diff
@@ -29,6 +29,7 @@
 import org.wso2.carbon.apimgt.api.dto.CertificateMetadataDTO;
 import org.wso2.carbon.apimgt.api.dto.ClientCertificateDTO;
 import org.wso2.carbon.apimgt.api.model.APIIdentifier;
+import org.wso2.carbon.apimgt.api.model.APIProductIdentifier;
```
Comment (vithu30) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java:32]: Seems this import is not used. Shall we remove it?

Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java:32]: This does not exist currently in [1].

[1] https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 11 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:5333]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9604/commits/08c2e6b56e8dd41ef907b3140d4601a985a2307d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManager.java (hunk 12 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManager.java:25]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9604/commits/77531e9f3feda9420f15f6cd639af122781038dd

**Pull request merged**